### DISCUSSION TITLE
User Profile: Value change listeners

### DISCUSSION THREAD

**Discussion by antikalk** [keycloak/keycloak#25379] [Ideas]
A nice thing to add to the User Profile Feature could be **listeners for value changes**. A possible scenario would e.g. be `phone_number` and `phone_number_verified`: Once the user changes his phone_number it would be nice to be able to add a listener to set `phone_number_verified` to false and require the user to verify it again.

These listeners would be called after the verification of the attribute value, but before it is persisted. Custom listeners could be provided by implementing a SPI.

**Comment by pedroigor** [comment]
What would be the scope of those listeners? Only notification about attribute changes or also changing their values before persisting?

For the former, we have an API for doing that but it would require writing your custom provider (and making this easier from a impl perspective). But if the goal is only about notifications when an attribute changes we could probably plug this into the server event system.

For the latter, we have this https://github.com/keycloak/keycloak/issues/23222.

**Reply by antikalk** [reply]
Cool, I like the proposal of https://github.com/keycloak/keycloak/issues/23222.

In my case I was more thinking about getting notified if a specific attribute value changes and then having the possibility to do something (e.g. add a required action or edit another user attribute that is only readonly for the user or admin). 

I think for those cases it might be enough to implement the UserProfile class and add our custom listeners there. Is there any documentation on how to configure Keycloak to then use this UserProfile Provider instead of the default one?

Of course it would be easier to extend Keycloak when having the possibility to inject custom Attribute Change Listeners into the DefaultUserProfile with an SPI. This could additionally be made configurable via the UI: you could add those change listeners to any user profile attribute via configuration (even multiple).

An interface for such a AttributeChangeListener SPI could look like this: 

```java
public interface AttributeChangeListenerSPI {

    /**
     * Returns the ID of the AttributeChangeListener.
     * This ID can be added to the User Profile Attribute metadata.
     * If an attribute value has been changed and it contains the ID of this
     * AttributeChangeListenerSPI in the list of configured change listeners
     * the onChange method is called.
     */
    String getId();

    /**
     * Method called for the attribute change.
     * 
     * @param name of the changed user attribute
     * @param context depending on the context you might wanna do different actions
     * @param user model where new attribute value is applied already (can be null if attribute is removed)
     * @param oldValue of the attribute before the change (can be null)
     */
    void onChange(String name, UserProfileContext context, UserModel user, List<String> oldValue);

}
```
